### DISCUSSION TITLE
To stress test network interest management by creating a large busy train station

### DISCUSSION THREAD

**Discussion by fire** [godotengine/godot-proposals#4233] [3D]
![image](https://user-images.githubusercontent.com/32321/158753246-4484701b-c7db-4598-903f-bddc6ace4b31.png)

This art for this demo would consist of a large train station model and some models of trains. It would consists on trains coming in and out of the station, and a huge amount (thousands) of agents (AI) entering and leaving trains all the time. This demo would allow to stress the following

Godot existing features:

- 3D Rendering a complex, realistic train station
- controlling thousands of agents pathfinding and AI
- testing local obstacle avoidance
- improving Navigation classes so they support threading
- rendering several thousand agents
- animating several thousand agents via compute shader with compressed animations
- using particle system collisions to deal with smoke inside the train station.

Game features:

- network interest management

Stretch goals:

- spatially dividing simulation over multiple servers

## Fales' comments

We can approximate the size of the state based on the size of the variants synchronized, and apply an average compression ratio
the profiler also shows you that bandwidth, but it's not smart, so it does not include things like protocol compression.

If you have a 1000 objects, each represented by 4 bytes ID + payload, you already have a 4K state, at 60Hz that's 240K/s and this is without payload (but also without compression).

ID Software's Quake 3 was 20 hertz tick rate (1998).

As a good actionable, we should test a few popular games bandwidth usage.

Start the client, measure the min/max/avg/deviation bandwidth in a real match scenario.

For example:

- Rocket League?
- COD?
- Battlefield?
- Fortnite?
- Dota 2?

## Fire's comments

What's a good estimate on a baseline channel rate on upload and download? 

1. of a player 
2. of a server

## References

* https://booth.pm/en/items/2878564
* Reduz's list of demo proposals on Google Drive

**Comment by HeadClot** [comment]
Hey just dropping [this](https://mirror-networking.gitbook.io/docs/interest-management/spatial-hashing) as it is a useful reference for Interest management.
